### ISSUE TITLE
LspPeekDefinition wrong position with OmniSharp

### ISSUE THREAD

**Issue by vkrajacic89** [prabirshrestha/vim-lsp#937]
When vim-lsp is connected to OmniSharp server, `LspPeekDefinition` alwalys show beginning of the file where definition is written, like its missing a line offset into that file. Its weird, because when I use `LspDefinition` command, it jumps to the right location.

This is the log file.
[vim-lsp.log](https://github.com/prabirshrestha/vim-lsp/files/5576267/vim-lsp.log)

This is my vim-lsp related config.
```
let g:lsp_diagnostics_enabled = 0
let g:lsp_signs_enabled = 0
let g:lsp_virtual_text_enabled = 0
let g:lsp_highlights_enabled = 0
let g:lsp_textprop_enabled = 0
let g:lsp_fold_enabled = 0
let g:lsp_highlight_references_enabled = 1
let g:lsp_highlight_references_delay = 100
let g:lsp_log_file = expand('~/vim-lsp.log')

if executable('OmniSharp')
	autocmd User lsp_setup call lsp#register_server({
		\ 'name': 'OmniSharp',
		\ 'cmd': ['OmniSharp', '-lsp'],
		\ 'root_uri':{server_info->lsp#utils#path_to_uri(
		\	lsp#utils#find_nearest_parent_file_directory(
		\		lsp#utils#get_buffer_path(),
		\		['.sln', '.svn/', '.git/']
		\	))},
		\ 'allowlist': ['cs'],
		\ })
endif

function! s:on_lsp_buffer_enabled() abort
	setlocal omnifunc=lsp#complete
	if exists('+tagfunc') | setlocal tagfunc=lsp#tagfunc | endif
endfunction
augroup lsp_install
	au!
	autocmd User lsp_buffer_enabled call s:on_lsp_buffer_enabled()
augroup END
```

**Comment by vkrajacic89** [comment]
I'm getting similar results with clangd in my C codebase. `LspPeekDefinition` most of the time shows bogus code. It opens right file but at wrong location. I wonder if I just have bad luck with language servers parsing my code or is this potential bug in plugn. Its a shame, because this is the only vim plugin I found that has peek functionality.

**Comment by thomasfaingnaert** [comment]
```
11/20/2020 10:51:48 PM:["s:on_stdout client request on_notification() error","Vim(let):E716: Key not present in Dictionary: linecount","function <SNR>53_out_cb[2]..<SNR>52_on_stdout[79]..<SNR>33_request_on_notification[3]..<SNR>34_createNext[1]..<SNR>34_subscribeSourceCallback[2]..<lambda>34[1]..<SNR>61_handle_location[39]..lsp#ui#vim#output#preview[45]..lsp#ui#vim#output#get_size_info, line 16"]
```

It seems your `getbufinfo()` doesn't return a `linecount`, which was added in Vim 8.2.0019. Updating Vim will likely resolve your issue.

**Comment by vkrajacic89** [comment]
I updated to latest nightly build. It kind "works" but it still showing wrong line.
When I inspect log, response I get from server seem to be pointing to right location.

`11/23/2020 9:28:17 AM:["<---",1,"OmniSharp",{"response":{"id":3,"jsonrpc":"2.0","result":{"uri":"file:///C%3A/Users/Vjekoslav/Projects/PineStudio/Buildings/Assets/Source/PineTools/PineTween/PineTween.cs","range":{"end":{"character":9,"line":242},"start":{"character":9,"line":242}}}},"request":{"id":3,"jsonrpc":"2.0","method":"textDocument/definition","params":{"textDocument":{"uri":"file:///C:/Users/Vjekoslav/Projects/PineStudio/Buildings/Assets/Source/Gameplay/Level.cs"},"position":{"character":22,"line":54}}}}]`

`range` param `start` and `end` are correct, line 242, but popup is showing line ~280 or somewhere near that.

**Comment by thomasfaingnaert** [comment]
Does `l:options['firstline']` at https://github.com/prabirshrestha/vim-lsp/blob/c2f216b9ab7e854e9b919f89d6115957dfb813d6/autoload/lsp/ui/vim/output.vim#L271 contain 242?

By the way, with #939 merged, you shouldn't need the nightly anymore.

**Comment by vkrajacic89** [comment]
`l:options['firstline']` is 238, but that might be because of default `center` peek alignment?

**Comment by thomasfaingnaert** [comment]
Yes, that sounds about right.
So somehow the `firstline` is not used correctly by Vim. Might be a problem when the text in the popup wraps?

**Comment by vkrajacic89** [comment]
I don't know, but in this concrete example, the range of lines [242, 280] only have 5 wrapped lines. And the error is much higher than 5.

**Comment by vkrajacic89** [comment]
I also tried fullscreening vim window and setting `g:lsp_preview_max_width` to huge value so wrapping doesn't occur. But the error is still present.

**Comment by vkrajacic89** [comment]
Ok, after hours of debugging (which is plain dum in the vimscript) I figured out its the problem in the data that is filled to the popup. Instead of filling it with text returned from the server, I just filled it with local file text and it works. 

So in `lsp#ui#vim#output#preview` function after the line
`let l:ft = lsp#ui#vim#output#append(a:data, l:lines, l:syntax_lines)`
I just inserted 
`let l:lines = a:data`
to override with local data and it magically works. I don't know exactly why is text returned from server different. Maybe the plugin author should see if returned data from the server should be parsed in some way or just use local file data like I did.
